packages/kbn-dev-utils/src/extract.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Fs from 'fs/promises';
+import { createWriteStream } from 'fs';
+import Path from 'path';
+import { pipeline } from 'stream';
+import { promisify } from 'util';
+
+import { lastValueFrom } from '@kbn/std';
+import Tar from 'tar';
+import Yauzl, { ZipFile, Entry } from 'yauzl';
+import * as Rx from 'rxjs';
+import { map, mergeMap, takeUntil } from 'rxjs/operators';
+
+const asyncPipeline = promisify(pipeline);
+
+interface Options {
+  /**
+   * Path to the archive to extract, .tar, .tar.gz, and .zip archives are supported
+   */
+  archivePath: string;
+
+  /**
+   * Directory where the contents of the archive will be written. Existing files in that
+   * directory will be overwritten. If the directory doesn't exist it will be created.
+   */
+  targetDir: string;
+
+  /**
+   * Number of path segments to strip form paths in the archive, like --strip-components from tar
+   */
+  stripComponents?: number;
+}
+
+/**
+ * Extract tar and zip archives using a single function, supporting stripComponents
+ * for both archive types, only tested with familiar archives we create so might not
+ * support some weird exotic zip features we don't use in our own snapshot/build tooling
+ */
+export async function extract({ archivePath, targetDir, stripComponents = 0 }: Options) {
```

Comment:
This functionality was already implemented in `@kbn/es` but not available in a shared location, though it feels like a great candidate for a shared util. Improved the implementation and error handling by using RxJS instead of manual promise and error handling.